File: backend/BearShop.Api/Models/ChatMessage.cs
namespace BearShop.Api.Models;

/// <summary>Một tin nhắn trong hội thoại giữa 1 khách hàng và shop (Admin).</summary>
public class ChatMessage
{
    public int Id { get; set; }

    /// <summary>Hội thoại thuộc về khách hàng nào (dù người gửi là Customer hay Admin).</summary>
    public int CustomerId { get; set; }
    public User? Customer { get; set; }

    /// <summary>"Customer" hoặc "Admin" — ai là người gửi tin này.</summary>
    public string SenderRole { get; set; } = "Customer";

    public string Text { get; set; } = string.Empty;
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
}

File: admin_web/lib/models/chat_message.dart
/// Một tin nhắn trong hội thoại chat (khách hàng <-> shop), góc nhìn Admin.
class ChatMessage {
  final String text;
  final bool fromAdmin;
  final DateTime time;

  const ChatMessage({
    required this.text,
    required this.fromAdmin,
    required this.time,
  });

  factory ChatMessage.fromJson(Map<String, dynamic> json) => ChatMessage(
    text: json['text'] ?? '',
    fromAdmin: (json['senderRole'] ?? 'Customer') == 'Admin',
    time:
        DateTime.tryParse(json['createdAt'] ?? '')?.toLocal() ??
        DateTime.now(),
  );
}

/// Một hội thoại (theo khách hàng) hiển thị trong danh sách chat của Admin.
class Conversation {
  final int customerId;
  final String customerName;
  final String lastMessage;
  final DateTime lastMessageAt;

  const Conversation({
    required this.customerId,
    required this.customerName,
    required this.lastMessage,
    required this.lastMessageAt,
  });

  factory Conversation.fromJson(Map<String, dynamic> json) => Conversation(
    customerId: json['customerId'],
    customerName: json['customerName'] ?? '',
    lastMessage: json['lastMessage'] ?? '',
    lastMessageAt:
        DateTime.tryParse(json['lastMessageAt'] ?? '')?.toLocal() ??
        DateTime.now(),
  );
}
